.github/workflows/build-and-deploy.yml

```diff
@@ -0,0 +1,153 @@
+    paths:
+      - .github/workflows/**
+      - e2eTests/**
+      - tests/**
+      - todo_app/**
+      - Dockerfile
+      - poetry.lock
+
+jobs:
+  test:
+    name: Build and test
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v4
+
+    - name: Set up Docker Buildx
+      uses: docker/setup-buildx-action@v3
+
+    - name: Cache Docker layers
+      uses: actions/cache@v4
+      with:
+        path: /tmp/.buildx-cache
+        key: ${{ runner.os }}-buildx-${{ github.sha }}
+        restore-keys: |
+          ${{ runner.os }}-buildx-   
+
+    - name: Integration test build
+      uses: docker/build-push-action@v6
+      with:
+        context: .
+        target: pipeline_integration_tests
+        tags: todo-app:pipeline_integration_tests
+        cache-from: type=local,src=/tmp/.buildx-cache
+        cache-to: type=local,dest=/tmp/.buildx-cache-new,mode=max
+        github-token: ${{ secrets.CACHE_GITHUB_TOKEN }}
+        load: true
+
+    - name: Run integration tests
+      run: docker run todo-app:pipeline_integration_tests
+
+    - name: e2e tests build
+      uses: docker/build-push-action@v6
+      with:
+        context: .
+        target: pipeline_e2e_tests
+        tags: todo-app:pipeline_e2e_tests
+        cache-from: type=local,src=/tmp/.buildx-cache
+        cache-to: type=local,dest=/tmp/.buildx-cache-new,mode=max
+        github-token: ${{ secrets.CACHE_GITHUB_TOKEN }}
+        load: true
+
+    - name: Move cache
+      run: |
+        rm -rf /tmp/.buildx-cache
+        mv /tmp/.buildx-cache-new /tmp/.buildx-cache      
+
+    - name: Run e2e tests
+      run: >
+        docker run
+        -e FLASK_APP=todo_app/app
+        -e TRELLO_TO_DO_BOARD_ID=65cb37a378d7b6f10d62c208
+        -e TRELLO_ORGANIZATION_ID=65cb37900e446a315b3b1fbc
+        -e TRELLO_API_KEY='TEST_API_KEY'
+        -e TRELLO_API_TOKEN='TEST_API_TOKEN'
+        todo-app:pipeline_e2e_tests
+
+    - name: Check for vulnerabilities
+      uses: snyk/actions/python-3.10@master
+      env:
+        SNYK_TOKEN: ${{ secrets.SNYK_TOKEN }}
+
+    - name: Report pipeline status
+      if: always()
+      id: slack
+      uses: slackapi/slack-github-action@v1.26.0
+      with:
+        payload: |
+          {
+            "name": ${{ toJson(format('Build and test: {0}', github.event.head_commit.message)) }},
+            "status": "${{ job.status }}"
+          }
+      env:
+        SLACK_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK }}
+
+  deploy:
+    name: Build and deploy
+    needs: test
+    if:  github.event_name == 'push' && github.ref == 'refs/heads/exercise-9'
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
+
+      - name: Cache Docker layers
+        uses: actions/cache@v4
+        with:
+          path: /tmp/.buildx-cache
+          key: ${{ runner.os }}-buildx-${{ github.sha }}
```

Comment:
Having a cache key seems somewhat irrelevant given we're clearing the old cache at the end of each run (admittedly I'm just critiquing the example from the Docker docs here).